File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let private nextNumber (n: int): int =
    match n % 2 with
    | 0 -> n / 2
    | 1 -> 3 * n + 1

let rec private countSteps (n: int, i: int): int * int =
    match n with
    | 1 -> (1, i)
    | _ -> countSteps (nextNumber n, i + 1)

let steps (number: int): int option =
    match number < 1 with
    | true -> None
    | false ->
        (number, 1)
        |> countSteps
        |> snd
        |> Some

File: solutions/scala/collatz-conjecture/1/src/main/scala/CollatzConjecture.scala
object CollatzConjecture
{

    private def nextNumber (n: Integer): Integer =
    {
        n % 2 match
        {
            case 0 => n / 2
            case 1 => 3 * n + 1
        }
    }

    private def countSteps (n: Integer, i: Integer): (Integer, Integer) =
    {
        n == 1 match
        {
            case true => (1, i)
            case false => countSteps(nextNumber(n), i + 1)
        }
    }

    def steps (number: Integer): Option[Integer] =
    {
        number < 1 match
        {
            case true => None
            case false => Some(countSteps(number, 0)._2)
        }
    }
}